### ISSUE TITLE
Let networks accept `label` keyword arguments

### ISSUE THREAD

**Issue by Seanny123** [nengo/nengo#745]
At the moment, the label of a network can't be set during it's initialization. For example,

```
prod_net = Product()
prod_net.label = 'pants'
```

This leaves code looking messy, however I don't know how complicated this would be to solve.

**Comment by tbekolay** [comment]
Yeah... this is kind of a weird one, but it is possible to set. Here's how:

```python
prod_net = Product(net=nengo.Network(label='pants'))
```

The intention here is to have only the product specific args in that constructor. Then if you need to set anything special on the network, you do it by passing in a pre-built network.

Thoughts? Needs more documentation / examples? Or is this too weird?

**Comment by Seanny123** [comment]
My thought is that this is way too weird and goes against the precedent set
by most other nengo objects.

**Comment by jgosmann** [comment]
Seems to weird to me. I think, we should keep arguments out of the constructor if they are passed to subcomponents of the network, but supporting the default network arguments seems fine.

**Comment by tcstewar** [comment]
> The intention here is to have only the product specific args in that constructor.

I'm fairly okay with arguments in the constructor that get passed on through to the base classes, if there's no weird transformation on those values.  For example, ```EnsembleArray``` takes a ```n_neurons``` argument, which it just passes on to ```Ensemble```.  But we don't want to do ```radius``` that way because it's got to have some wacky ```1/dimensions``` scaling applied to it.  So I think custom ```Networks``` should be allowed to set parameters that have a direct mapping.  

The only difficulty is how we define those arguments.  I'd be tempted to *not* use ```**kwargs``` and rather be explicit and just have a small subset of common arguments (```label``` is the only one coming to mind).

**Comment by bjkomer** [comment]
That syntax feels really weird to me. I think that's because I was thinking that these networks were subclasses of Network, but they are actually just functions with capital letters. Knowing that, I can see why this syntax would make sense. It will also allow people to be able to do funky things like:
```
prod_net = Product(net=nengo.Oscillator())
```
Which might be useful in some cases. I still think it would be useful to be able to specify things common to all networks such as `label`, and maybe `seed` in the constructor for custom `Networks`.

**Comment by Seanny123** [comment]
Alright, I'll add `label` and `seed` as arguments common to all `Networks`. Anything else?

**Comment by hunse** [comment]
If someone does `prod = Product(net=nengo.Network(), seed=3, label='prod')`, do we want to set (or overwrite) the seed and label on their passed-in network, or throw an error that the "net" parameter is mutually exclusive with "seed" and "label"?

**Comment by Seanny123** [comment]
I'd go for over-write and warn.

**Comment by tbekolay** [comment]
Hmm... it feels weird to me to duplicate these things. Especially since there's a third parameters, `add_to_container`... and what if later on there becomes more parameters? I agree that it's suboptimal having two ways to make networks (functions vs. subclasses) but I think it'd be better to figure out a way to make these two situations more visually distinct than have two ways to do the same thing...  my two cents anyhow!

**Comment by Seanny123** [comment]
What do you mean by "visually distinct"?

**Comment by tbekolay** [comment]
Like, you should be able to tell by looking at the line of code making the network whether it's a Network subclass or not. The most obvious way to do this would be the not capitalize the first letter. E.g.,
```python
cconv = nengo.networks.circular_convolution()  # not a subclass (doesn't accept label)
ea = nengo.networks.EnsembleArray()  # a subclass (accepts label)
```

Also, in case anyone's curious why things are this way, I took this convention from Matplotlib. All `plt.*  plotting functions are just shortcuts to functions on the `axes` object. So, if you don't already have an `axes` object, Matplotlib will automatically make one for you, and call the appropriate plotting function on the `axes` object. In this way, you can easily make a simple plot with just a `plt.plot` call, or you can draw many things to the same set of `axes` by either doing `ax.plot` or `plt.plot(ax=ax)`.

Networks are kind of analogous to `axes`; all Nengo objects are contained in a network, as all plots are drawn on an `axes`. So, my idea was basically to allow you to create a new network simply (by not passing a network) or to add in this function's new objects to the network by passing one in. It also got around the cumbersome issue of having to duplicate the network's arguments (like `label`, `seed`) by having people pass in a pre-existing network if desired. Similarly, if you have a bunch of config stuff you want to set on a network, you can set that in the network's config object, and then pass it that network.

On the Matplotlib analogy, imagine you want to make a simple plot on a figure with a custom `figsize`. Rather than adding `figure` params to `plot`, you'd be instructed to just make your figure first. Here's the most explicit version of this.
```python
fig = plt.figure(figsize=(5, 2))
ax = fig.add_subplot(111)
ax.plot(x, y)
```
For convenience (or, perhaps, mostly to look like Matlab) you can instead do this
```python
fig = plt.figure(figsize=(5, 2))
ax = fig.add_subplot(111)
plt.plot(x, y, ax=ax)
```
And, as additional magic (which Matplotlib has, but Nengo decided against) there is this idea of the "current" figure and "current" axes that you, by default, draw on.
```python
fig = plt.figure(figsize=(5, 2))
plt.plot(x, y)  # Automatically makes a new `axes` object on the current figure, fig.
```

So yeah, that's why I did things that way. Of course, it stands to reason that, having not allowed that last possibility, maybe we shouldn't allow the second possibility either, and instead force people to call the network building functions inside a `with network:` block instead of doing some sneaky things inside those functions that may not be obvious at first glance.

**Comment by jgosmann** [comment]
I don't think, that the matplotlib analogy maps very well onto nengo. There doesn't seem to be a parallel to the first case which would something like
```python
net = nengo.Network(
net.create_product(...)
```

I never used the default matplotlib function in the way of the second possibility, so it didn't even come to my mind to consider the Nengo API in this way. But even if you do, there is some difference. If I call `plt.plot(..., ax=ax)`, it adds further objects, which might overlay other objects, but nothing existing is removed. In Nengo, however,
```python
Product(net=net)
CircularConvolution(net=net)
```
would overwrite objects (both assign `net.A`, `net.B`, and `net.output`).

I always thought of the `plt` plotting functions as using the current default axes object, but in Nengo we don't have such a default object.


I prefer to treat the network creation functions as returning some object which I just connect up and I don't have to care about the internals. If I pass in a `net`, I somehow have to know what the functions changes in my network.

**Comment by tbekolay** [comment]
> would overwrite objects (both assign net.A, net.B, and net.output)

Technically, it would only overwrite the references `net.A`, `net.B` and `net.output`, not the actual objects; all of the objects will still exist and be simulated. So this is ok

```python
Product(net=net)
nengo.Connection(prod_a, net.A)
nengo.Connection(prob_b, net.B)
CircularConvolution(net=net)
nengo.Connection(cconv_a, net.A)
nengo.Connection(cconv_b, net.B)
```

but this is not

```python
Product(net=net)
CircularConvolution(net=net)
nengo.Connection(prod_a, net.A)
nengo.Connection(prob_b, net.B)
nengo.Connection(cconv_a, net.A)
nengo.Connection(cconv_b, net.B)
```

> I always thought of the plt plotting functions as using the current default axes object, but in Nengo we don't have such a default object.

We don't anymore! But we used to... I remember those days still. Not that we should go back to doing it, but for the context of my mental model of the API at the time.

**Comment by Seanny123** [comment]
Going back to visual distinctiveness, what exactly did you have in mind for this case of labels and additional keywords? Or are you still asking for suggestions? Should we bring this up in the next Nengo dev meeting (which I unfortunately won't be able to attend).

**Comment by jgosmann** [comment]
This should be resolved by #1179.
